File: web/src/lib/providers/gemini.ts
import { GoogleGenAI } from "@google/genai";
import { LENGTH_PRESETS, MODEL_MAP, type Length, type Speed } from "../config";
import type { ProviderResponse } from "./types";

const ai = new GoogleGenAI({ apiKey: process.env.GOOGLE_API_KEY ?? "" });

/** Minimum thinking budget for gemini-2.5-pro (thinking can't be disabled). */
const THINKING_BUDGET = 128;

export async function queryGemini(
  prompt: string,
  speed: Speed,
  length: Length,
  { webSearch = true }: { webSearch?: boolean } = {}
): Promise<ProviderResponse> {
  const modelName = MODEL_MAP.gemini[speed];
  const preset = LENGTH_PRESETS[length];
  const isThinkingModel = modelName.includes("2.5");

  // When web search (Google Search grounding) is enabled the model needs
  // enough output-token headroom for grounding chunks + the answer itself.
  // Mirror the Claude approach: floor at 4096 when search is on.
  const baseTokens = webSearch
    ? Math.max(preset.maxTokens, 4096)
    : preset.maxTokens;
  const maxOutputTokens = isThinkingModel
    ? baseTokens + THINKING_BUDGET
    : baseTokens;

  const t0 = performance.now();
  const result = await ai.models.generateContent({
    model: modelName,
    contents: prompt,
    config: {
      maxOutputTokens,
      systemInstruction: preset.systemInstruction,
      ...(isThinkingModel && {
        thinkingConfig: { thinkingBudget: THINKING_BUDGET },
      }),
      ...(webSearch && { tools: [{ googleSearch: {} }] }),
    },
  });
  const latency = (performance.now() - t0) / 1000;

  const text = result.text ?? "";
  const usage = result.usageMetadata;

  return {
    engine: "gemini",
    model: modelName,
    text,
    latencySeconds: Math.round(latency * 100) / 100,
    inputTokens: usage?.promptTokenCount ?? 0,
    outputTokens: usage?.candidatesTokenCount ?? 0,
  };
}

File: metallm/providers/gemini.py
"""Google Generative AI / Gemini provider adapter."""

from __future__ import annotations

import os

from google import genai
from google.genai import types

from metallm.config import LENGTH_PRESETS, MODEL_MAP, Engine, Length, Speed
from metallm.providers.base import BaseProvider, ProviderResponse


class GeminiProvider(BaseProvider):
    engine = Engine.GEMINI

    def __init__(self) -> None:
        api_key = os.environ.get("GOOGLE_API_KEY", "")
        self._client = genai.Client(api_key=api_key)

    async def query(self, prompt: str, speed: Speed, length: Length) -> ProviderResponse:
        model = MODEL_MAP[self.engine][speed]
        preset = LENGTH_PRESETS[length]

        config = types.GenerateContentConfig(
            system_instruction=preset.system_instruction,
            max_output_tokens=preset.max_tokens,
        )

        t0 = self._timer()
        response = await self._client.aio.models.generate_content(
            model=model,
            contents=prompt,
            config=config,
        )
        latency = self._timer() - t0

        text = response.text or ""
        usage_meta = response.usage_metadata

        return ProviderResponse(
            engine=self.engine,
            model=model,
            text=text,
            latency_seconds=round(latency, 2),
            input_tokens=getattr(usage_meta, "prompt_token_count", 0) or 0,
            output_tokens=getattr(usage_meta, "candidates_token_count", 0) or 0,
        )
